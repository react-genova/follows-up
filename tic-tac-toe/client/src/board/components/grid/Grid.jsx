import React from 'react';
import PropTypes from 'prop-types';
import { GridContainer } from './Grid.styled';
import { playingTypePropTypes, valuesPropTypes } from './Grid.types';
import GridCell from './GridCell';

const getIndex = index => index;

const Grid = ({ values, playingType, onCellClick }) => (
    <GridContainer>
        {
            values.map(({ value, valid }, index) => (
                <GridCell
                    key={getIndex(index)}
                    valid={valid}
                    value={value}
                    index={index}
                    playingType={playingType}
                    onClick={onCellClick}
                />
            ))
        }
    </GridContainer>
);

Grid.propTypes = {
    values: valuesPropTypes.isRequired,
    playingType: playingTypePropTypes.isRequired,
    onCellClick: PropTypes.func,
};

Grid.defaultProps = {
    onCellClick: undefined,
};

export default Grid;
