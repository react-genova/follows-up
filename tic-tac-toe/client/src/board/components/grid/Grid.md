### Playing X on empty little board

```jsx
const val = value => ({ value, valid: SIGN_NONE !== value });
const { SIGN_O, SIGN_X, SIGN_NONE } = require('../../modules/types/signs.constants.js');
const values = [SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE].map(val);
<div style={{ width:"160px", height:"160px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Grid values={values} playingType={SIGN_X} highlightSequence={[]} />
</div>
```

### Playing O on empty medium board

```jsx
const val = value => ({ value, valid: SIGN_NONE !== value });
const { SIGN_O, SIGN_X, SIGN_NONE } = require('../../modules/types/signs.constants.js');
const values = [SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_NONE].map(val);
<div style={{ width:"240px", height:"240px", padding:"10px", backgroundColor: "rgb(50,50,50)" }}>
    <Grid values={values} playingType={SIGN_O} highlightSequence={[]} />
</div>
```

### Playing X on board

```jsx
const val = value => ({ value, valid: SIGN_NONE !== value });
const { SIGN_O, SIGN_X, SIGN_NONE } = require('../../modules/types/signs.constants.js');
const values = [SIGN_X, SIGN_X, SIGN_O, SIGN_O, SIGN_NONE, SIGN_NONE, SIGN_NONE, SIGN_X, SIGN_O].map(val);
<div style={{ width:"320px", height:"320px", padding:"10px", backgroundColor: "rgb(50,50,50)" }}>
    <Grid values={values} playingType={SIGN_X} highlightSequence={[]} />
</div>
```

### Playing O on board

```jsx
const val = value => ({ value, valid: SIGN_NONE !== value });
const { SIGN_O, SIGN_X, SIGN_NONE } = require('../../modules/types/signs.constants.js');
const values = [SIGN_NONE, SIGN_O, SIGN_X, SIGN_X, SIGN_NONE, SIGN_O, SIGN_NONE, SIGN_X, SIGN_NONE].map(val);
<div style={{ width:"320px", height:"320px", padding:"10px", backgroundColor: "rgb(50,50,50)" }}>
    <Grid values={values} playingType={SIGN_O} highlightSequence={[]} />
</div>
```

### Highlighting values

```jsx
const val = value => ({ value, valid: SIGN_NONE !== value });
const { SIGN_O, SIGN_X, SIGN_NONE } = require('../../modules/types/signs.constants.js');
const values = [SIGN_X, SIGN_X, SIGN_X, SIGN_O, SIGN_NONE, SIGN_O, SIGN_NONE, SIGN_O, SIGN_NONE].map(val);
<div style={{ width:"320px", height:"320px", padding:"10px", backgroundColor: "rgb(50,50,50)" }}>
    <Grid values={values} playingType={SIGN_O} highlightSequence={[0, 1, 2, 3, 5, 7]} />
</div>
```
