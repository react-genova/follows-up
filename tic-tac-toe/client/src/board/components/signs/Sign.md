### With values

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
</div>
```

### Without values (hover)

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```

### A little bigger and mixed

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"480px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```