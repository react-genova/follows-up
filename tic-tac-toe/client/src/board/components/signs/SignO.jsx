import React from 'react';
import PropTypes from 'prop-types';

const SignO = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <ellipse cx="50" cy="50" rx="46" ry="46" fill="transparent" stroke={ghost ? 'lightgray' : 'yellow'} strokeWidth="8" />
    </svg>
);

SignO.propTypes = {
    ghost: PropTypes.bool,
};

SignO.defaultProps = {
    ghost: false,
};

export default SignO;
