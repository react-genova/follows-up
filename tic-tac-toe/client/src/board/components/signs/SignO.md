### On light background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px" }}>
    <SignO />
    <SignO ghost />
</div>
```

### On dark background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```

### A little bigger

```jsx
<div style={{ display:"flex", width:"200px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```