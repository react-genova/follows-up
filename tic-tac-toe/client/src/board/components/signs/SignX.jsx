import React from 'react';
import PropTypes from 'prop-types';

const SignX = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <line x1="0" y1="0" x2="100" y2="100" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
        <line x1="0" y1="100" x2="100" y2="0" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
    </svg>
);

SignX.propTypes = {
    ghost: PropTypes.bool,
};

SignX.defaultProps = {
    ghost: false,
};


export default SignX;
