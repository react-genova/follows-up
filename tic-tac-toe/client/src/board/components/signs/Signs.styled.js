import styled from 'styled-components';

const SignContainer = styled.div`
    width: 100%;
    height: 100%;
`;

export default SignContainer;
