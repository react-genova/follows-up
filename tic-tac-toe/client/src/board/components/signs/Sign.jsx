import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
import * as SIGNS from '../../modules/types/signs.constants';
import SignContainer from './Signs.styled';
import SignO from './SignO';
import SignX from './SignX';
import { playingTypePropTypes } from '../grid/Grid.types';

const Sign = ({
    className, hasValue, onClick, type,
}) => {
    const [hover, setHover] = useState(false);
    return (
        <SignContainer
            className={className}
            onClick={onClick}
            onMouseEnter={() => setHover(true)}
            onMouseLeave={() => setHover(false)}
        >
            {
                (hasValue || hover) && {
                    [SIGNS.SIGN_O]: <SignO ghost={!hasValue} />,
                    [SIGNS.SIGN_X]: <SignX ghost={!hasValue} />,
                }[type]
            }
        </SignContainer>
    );
};

Sign.propTypes = {
    className: PropTypes.string,
    hasValue: PropTypes.bool,
    onClick: PropTypes.func,
    type: playingTypePropTypes,
};

Sign.defaultProps = {
    className: undefined,
    hasValue: false,
    onClick: undefined,
    type: SIGNS.SIGN_O,
};

export default memo(Sign);
