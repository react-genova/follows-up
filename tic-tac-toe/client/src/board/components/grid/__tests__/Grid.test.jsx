import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import Grid from '../Grid';
import { SIGN_O, SIGN_X, SIGN_NONE } from '../../../modules/types/signs.constants';

describe('Grid component', () => {
    afterEach(cleanup);

    it('renders without crashing', () => {
        const VALUES = [
            { value: SIGN_O, valid: true },
            { value: SIGN_X, valid: true },
            { value: SIGN_NONE, valid: true },
        ];
        const { container } = render(<Grid values={VALUES} playingType={SIGN_O} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
