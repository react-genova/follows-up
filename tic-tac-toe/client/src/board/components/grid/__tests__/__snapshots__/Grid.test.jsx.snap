// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component renders without crashing empty hightlight sequence 1`] = `
<div
  class="sc-bdVaJa cpgzAv"
>
  <div
    class="sc-bwzfXH evblgu"
    type="MOVING_SYMBOL_O"
  >
    <div
      class="cell-0 sc-htpNat emEaka"
    >
      <svg
        class="sign-o"
        height="100%"
        viewBox="0 0 100 100"
        width="100%"
      >
        <ellipse
          cx="50"
          cy="50"
          fill="transparent"
          rx="46"
          ry="46"
          stroke="yellow"
          stroke-width="8"
        />
      </svg>
    </div>
  </div>
  <div
    class="sc-bwzfXH evblgu"
    type="MOVING_SYMBOL_X"
  >
    <div
      class="cell-1 sc-htpNat emEaka"
    >
      <svg
        class="sign-x"
        height="100%"
        viewBox="0 0 100 100"
        width="100%"
      >
        <line
          stroke="cyan"
          stroke-width="8"
          x1="0"
          x2="100"
          y1="0"
          y2="100"
        />
        <line
          stroke="cyan"
          stroke-width="8"
          x1="0"
          x2="100"
          y1="100"
          y2="0"
        />
      </svg>
    </div>
  </div>
  <div
    class="sc-bwzfXH evblgu"
    type="NONE"
  >
    <div
      class="cell-2 sc-htpNat emEaka"
    />
  </div>
</div>
`;

exports[`Grid component renders without crashing valid winning sequence 1`] = `
<div
  class="sc-bdVaJa cpgzAv"
>
  <div
    class="sc-bwzfXH evblgu"
    type="MOVING_SYMBOL_O"
  >
    <div
      class="cell-0 sc-htpNat emEaka"
    >
      <svg
        class="sign-o"
        height="100%"
        viewBox="0 0 100 100"
        width="100%"
      >
        <ellipse
          cx="50"
          cy="50"
          fill="transparent"
          rx="46"
          ry="46"
          stroke="yellow"
          stroke-width="8"
        />
      </svg>
    </div>
  </div>
  <div
    class="sc-bwzfXH haqnfV"
    type="MOVING_SYMBOL_X"
  >
    <div
      class="cell-1 sc-htpNat emEaka"
    >
      <svg
        class="sign-x"
        height="100%"
        viewBox="0 0 100 100"
        width="100%"
      >
        <line
          stroke="cyan"
          stroke-width="8"
          x1="0"
          x2="100"
          y1="0"
          y2="100"
        />
        <line
          stroke="cyan"
          stroke-width="8"
          x1="0"
          x2="100"
          y1="100"
          y2="0"
        />
      </svg>
    </div>
  </div>
  <div
    class="sc-bwzfXH evblgu"
    type="NONE"
  >
    <div
      class="cell-2 sc-htpNat emEaka"
    />
  </div>
</div>
`;
